design/backup-performance-improvements.md

```diff
@@ -0,0 +1,334 @@
+- Once ItemBlock is determined, call new func `BackupItemBlock` instead of `BackupItem`.
+- New func `BackupItemBlock` will call pre hooks for any pods in the block, then back up the items in the block (`BackupItem` will no longer run hooks directly), then call post hooks for any pods in the block.
+
+### Phase 2: Process ItemBlocks for a single backup in multiple threads
+- Concurrent `BackupItemBlock` operations will be executed by worker threads invoked by the backup controller, which will communicate with the backup controller operation via a shared channel.
+- The ItemBlock processing loop implemented in Phase 1 will be modified to send each newly-created ItemBlock to the shared channel rather than calling `BackupItemBlock` inline.
+- Users will be able to configure the number of workers available for concurrent `BackupItemBlock` operations.
+- Access to the BackedUpItems map must be synchronized
+
+## Detailed Design
+
+### Phase 1: ItemBlock processing
+
+#### New ItemBlockAction plugin type
+
+In order for Velero to identify groups of items to back up together in an ItemBlock, we need a way to identify items which need to be backed up along with the current item. While the current `Execute` BackupItemAction method does return a list of additional items which are required by the current item, we need to know this *before* we start the item backup. To support this, we need a new plugin type, `ItemBlockAction` (IBA) with an API method, `GetRelatedItems` which Velero will call on each item as it processes. The expectation is that the registered IBA plugins will return the same items as returned as additional items by the BIA `Execute` method, with the exception that items which are not created until calling `Execute` should not be returned here, as they don't exist yet.
+
+#### Proto definition (compiled into golang by protoc)
+
+The ItemBlockAction plugin type is defined as follows:
+```
+service ItemBlockAction {
+    rpc AppliesTo(ItemBlockActionAppliesToRequest) returns (ItemBlockActionAppliesToResponse);
+    rpc GetRelatedItems(ItemBlockActionGetRelatedItemsRequest) returns (ItemBlockActionGetRelatedItemsResponse);
+}
+
+message ItemBlockActionAppliesToRequest {
+    string plugin = 1;
+}
+
+message ItemBlockActionAppliesToResponse {
+    ResourceSelector ResourceSelector = 1;
+}
+
+message ItemBlockActionRelatedItemsRequest {
+    string plugin = 1;
+    bytes item = 2;
+    bytes backup = 3;
+}
+
+message ItemBlockActionRelatedItemsResponse {
+    repeated generated.ResourceIdentifier relatedItems = 1;
+}
+```
+
+A new PluginKind, `ItemBlockActionV1`, will be created, and the backup process will be modified to use this plugin kind.
+
+For any BIA plugins which return additional items from `Execute()` that need to be backed up at the same time or sequentially in the same worker thread as the current items should add a new IBA plugin to return these same items (minus any which won't exist before BIA `Execute()` is called).
```

Comment:
This is needed because this is the only way we make sure we guarantee that the items get backed up in the same worker thread at the same time. If we don't have an IBA action which adds these items to the item block, then all of the items will be considered as belonging to separate ItemBlocks. In some scenarios things will work just fine, but in others they may not.

As an example, lets take a backup with Pods, PVCs, and PVs and nothing else -- each Pod contains one mounted PVC bound to a PV. If we create IBA plugins for Pod and PVC, then when we create ItemBlocks, each block will contain one pod, one PVC, and one PV. If there's just one pod, there's one item block. If there are 100 pods, there are 100 ItemBlocks. If we don't create the new IBA plugins for Pod and PVC, then Velero will just have isolated ItemBlocks with Pods, then PVCs, then PVs, since the IBA action is what groups items into blocks.

Now lets look at two scenarios -- the first will work as you expect (additional items will be handled within BackupItemBlock), and in the second they may not.

Scenario 1: Number of pods is much larger than number of ItemBlock workers.
In this case, say we have 100 pods and 5 workers. While the pre-calculated ItemBlocks will only have one item each, the Pod blocks will be first, followed by the PVC blocks, then the PV blocks. So the first 5 pod blocks get sent to the 5 workers. As each is processed, the Execute() call pulls in the additional items, so we get each pod backed up with its PVC and PV, and all 3 items are marked as already backed up. As one finishes, another Pod block is pulled off the queue. By the time the first PVC block gets pulled in, 96 out of 100 pods (along with their PVCs) have already been backed up, so this PVC has a 96% chance of already being backed up -- the first thing Velero does in BackupItem is to see whether this item has already been backed up, and if it has, it exits immediately. The block is done, pull the next, etc. It's possible that all 100 PVCs will have already been handled before their PVC blocks arrive in the worker pool -- at most, 4 of these may not have, in which case PVC backup and pod hooks may run out-of-order -- which is the risk when we don't have an IBA plugin defined if there are interdependencies.

Scenario 2: There are many more worker threads than pods
In this case, lets assume there are only 2 pods but 10 workers. In this case, both pods, both PVCs, and both PVs are all sent to workers at the same time. So the likelihood is high that by the time the Pod plugin gets that PVC additional item return, it's already been started backing up in another worker, so it will be skipped here. This means, especially if it's a large volume, that the post hook is likely to run *before* the PVC snapshot has been created.

Creating the IBA plugin eliminates this problem entirely. This is the whole reason we have the IBA plugin -- it enables us to create ItemBlocks based on item interdependencies. If we don't have this, then we can't create meaningful ItemBlocks.

Reply:
Thanks for the explanation.  I didn't consider "scenario 2".  
But in regards to the question of whether a BIA plugin returning `additionalItems` should have an IBA to group these items earlier, IMO the answer is till `That depends`.  B/C sometimes the additional items do not have to be in the same `itemBlock` with the "original item".  Another possibility is that some of these additional items may be created within the `Execute()` function of BIA.

Reply:
@reasonerjt "Another possibility is that some of these additional items may be created within the Execute() function of BIA." -- yes, this is called out explicitly as a case where the items will not be in the ItemBlock. Actually, your "that depends" is also covered in the text that we're responding to now. "For any BIA plugins which return additional items from `Execute()` that need to be backed up at the same time or sequentially in the same worker thread as the current items should add a new IBA plugin to return these same items" -- the key part here is "that need to be backed up at the same time or sequentially in the same worker thread" -- so "sometimes the additional items do not have to be backed up in the  same ItemBlock with the original item", in that case they do not need to be in the IBA. What you said in your comment and what's in the design say the same thing, just in different words.

Reply:
OK, sounds good to me.